GraalVM Native Image Requires Config for lazy vals

## Compiler version

`3.1.0`

## Background

GraalVM Native Image Ahead-of-Time compiles JVM apps into native executables.  To do the AOT compilation, reflection points must be specified.

## Issue 1

Scala's `scala.runtime.LazyVals` uses reflection so it must be declared.  This could be done in the Scala stdlib so that others don't need to do it themselves.  But this approach is a bit brittle.  Any chance of removing the use of `sun.misc.Unsafe` there?  If not, then I can send a PR to include the GraalVM reflection config:
```
  {
    "name":"sun.misc.Unsafe",
    "allDeclaredFields":true
  }
```

## Issue 2

Code generated by the compiler for lazy vals also uses reflection and must be declared, like:
```
  {
    "name":"App$package$",
    "fields":[{"name":"0bitmap$1"}]
  }
```

Also, ideally reflection is not required.  But I haven't looked into what the reflection is doing.  If someone uses a Scala library that uses lazy vals, and does NOT provide their reflection config for GraalVM, then the user must specify the config for the library on their own.  Ideally the library authors nor the users need to specify this info.

## Steps to Reproduce

1. Clone `https://github.com/jamesward/scala3-graalvm-tester.git`
1. Switch to the `lazy-reflect-broke` branch
1. Run `./sbt nativeImage nativeImageRun`

You will see a runtime error showing Issue 1:
```
Exception in thread "main" scala.runtime.LazyVals$$anon$1
	at scala.runtime.LazyVals$.$init$$$anonfun$3(LazyVals.scala:18)
	at scala.Option.getOrElse(Option.scala:201)
	at scala.runtime.LazyVals$.<clinit>(LazyVals.scala:19)
	at App$package$.<clinit>(App.scala:2)
	at run.main(App.scala:4)
```

(Note: It should be possible to fail the build in this case instead of producing a runtime error but there seems to be a bug: https://github.com/oracle/graal/issues/4010)

If you switch to the `lazy-reflect-works` branch and re-run `./sbt nativeImage nativeImageRun` you will see that it works given [the requisite config](https://github.com/jamesward/scala3-graalvm-tester/blob/lazy-reflect-works/src/main/resources/META-INF/native-image/reflect-config.json) for the app in that repo that has one lazy val.

## One Approach (If Reflection Must Stay)

The Spring team has developed a set of annotations which provide the reflection info, like:
https://docs.spring.io/spring-native/docs/0.9.1-SNAPSHOT/api/org/springframework/nativex/hint/TypeHint.html

There has been talk that this work will merge in some way into GraalVM itself.  In that case, the Scala compiler could generate the necessary `TypeHint` annotations for places it generates code that uses reflection.

I have a sample project using ZIO and Scala 3 and this issue happens. Running the Native Image binary gave me:

```
❯ /Users/cdepaula/repos/scala-playground/zio-scalajs-stack/out/backend/nativeImage.dest/backend
Exception in thread "main" scala.runtime.LazyVals$$anon$1
	at scala.runtime.LazyVals$.$init$$$anonfun$3(LazyVals.scala:18)
	at scala.Option.getOrElse(Option.scala:201)
	at scala.runtime.LazyVals$.<clinit>(LazyVals.scala:19)
	at zio.ZLayer.<clinit>(ZLayer.scala:42)
	at java.lang.Class.ensureInitialized(DynamicHub.java:525)
	at zio.ZLayer$Suspend$.apply(ZLayer.scala:437)
	at zio.ZLayer$.suspend(ZLayer.scala:667)
	at zio.ZLayer$.fromZIOEnvironment(ZLayer.scala:613)
	at zio.ZLayer$.<clinit>(ZLayer.scala:558)
	at zio.ZIOAppDefault.$init$(ZIOAppDefault.scala:43)
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:11)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
```

I tried @jamesward workaround by creating `/backend/resources/META-INF/native-image/reflect-config.json` with:

```json
[
  {
    "name":"sun.misc.Unsafe",
    "allDeclaredFields":true
  },
  {
    "name":"App$package$",
    "fields":[{"name":"0bitmap$1"}]
  }
]
```

and then had the error:

```
❯ /Users/cdepaula/repos/scala-playground/zio-scalajs-stack/out/backend/nativeImage.dest/backend
Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.lang.Class.ensureInitialized(DynamicHub.java:525)
	at zio.ZLayer$Suspend$.apply(ZLayer.scala:437)
	at zio.ZLayer$.suspend(ZLayer.scala:667)
	at zio.ZLayer$.fromZIOEnvironment(ZLayer.scala:613)
	at zio.ZLayer$.<clinit>(ZLayer.scala:558)
	at zio.ZIOAppDefault.$init$(ZIOAppDefault.scala:43)
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:11)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
Caused by: java.lang.NoSuchFieldException: 0bitmap$1
	at java.lang.Class.getDeclaredField(DynamicHub.java:968)
	at scala.runtime.LazyVals$.getOffset(LazyVals.scala:103)
	at zio.ZLayer.<clinit>(ZLayer.scala:42)
	... 8 more
```

Also tried removing the `bitmal$1` field but same happens. I'm using Scala 3.2.0. Anything I missed?

Hey @szymon-rd , thanks a lot for the PR. I tried to go as deep as possible on this since the sample from above (simple LazyVal) worked but my sample app in Zio didn't.

I went on adding the unsafe fields to `reflect-config.json` for each error I found but stopped since it went on and on. 

One thing in common is that all of them refers to the `getOffset` method from `LazyVals.scala` so seems some fix on it might fix it all.

Below the reflect-config:

```json
[
    {
        "name": "izumi.reflect.macrortti.LightTypeTag",
        "fields": [
            {
                "name": "0bitmap$1",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "izumi.reflect.macrortti.LightTypeTag$ParsedLightTypeTag210",
        "fields": [
            {
                "name": "0bitmap$5",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "zhttp.http.Method$PUT$",
        "fields": [
            {
                "name": "0bitmap$6",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "zhttp.http.Method$DELETE$",
        "fields": [
            {
                "name": "0bitmap$8",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "zhttp.http.Method$POST$",
        "fields": [
            {
                "name": "0bitmap$5",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "zhttp.http.Method$GET$",
        "fields": [
            {
                "name": "0bitmap$3",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "izumi.reflect.macrortti.LightTypeTagRef$NameReference",
        "fields": [
            {
                "name": "0bitmap$5",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "izumi.reflect.macrortti.LightTypeTagInheritance",
        "fields": [
            {
                "name": "0bitmap$1",
                "allowUnsafeAccess": true
            }
        ]
    },
    {
        "name": "izumi.reflect.internal.fundamentals.platform.console.TrivialLogger$Config",
        "fields": [
            {
                "name": "0bitmap$1",
                "allowUnsafeAccess": true
            }
        ]
    }
]
```

And the errors I found each execution:

```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.base@17.0.5/java.lang.Class.ensureInitialized(DynamicHub.java:528)
	at izumi.reflect.macrortti.LightTypeTag$.parse(LightTypeTag.scala:271)
	at zio.ZIOAppDefault.$init$(ZIOAppDefault.scala:45)
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:10)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
Caused by: java.lang.NoSuchFieldException: 0bitmap$1
	at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
	at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
	at izumi.reflect.macrortti.LightTypeTag.<clinit>(LightTypeTag.scala:49)
	... 5 more

--------------------------------------------------------------------------------
Exception in thread "main" java.lang.ExceptionInInitializerError
	at izumi.reflect.macrortti.LightTypeTag$.parse(LightTypeTag.scala:271)
	at zio.ZIOAppDefault.$init$(ZIOAppDefault.scala:45)
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:10)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
Caused by: java.lang.NoSuchFieldException: 0bitmap$5
	at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
	at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
	at izumi.reflect.macrortti.LightTypeTag$ParsedLightTypeTag210.<clinit>(LightTypeTag.scala:375)
	... 4 more

--------------------------------------------------------------------------------
Exception in thread "main" java.lang.ExceptionInInitializerError
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:15)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
Caused by: java.lang.NoSuchFieldException: 0bitmap$6
	at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
	at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
	at zhttp.http.Method$PUT$.<clinit>(Method.scala:58)
	... 2 more

--------------------------------------------------------------------------------
Exception in thread "main" java.lang.ExceptionInInitializerError
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:15)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
Caused by: java.lang.NoSuchFieldException: 0bitmap$8
	at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
	at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
	at zhttp.http.Method$DELETE$.<clinit>(Method.scala:60)
	... 2 more

--------------------------------------------------------------------------------
Exception in thread "main" java.lang.ExceptionInInitializerError
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:15)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
Caused by: java.lang.NoSuchFieldException: 0bitmap$5
	at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
	at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
	at zhttp.http.Method$POST$.<clinit>(Method.scala:57)
	... 2 more

--------------------------------------------------------------------------------
Exception in thread "main" java.lang.ExceptionInInitializerError
	at com.carlosedp.zioscalajs.backend.MainApp$.<clinit>(MainApp.scala:15)
	at com.carlosedp.zioscalajs.backend.MainApp.main(MainApp.scala)
Caused by: java.lang.NoSuchFieldException: 0bitmap$3
	at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
	at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
	at zhttp.http.Method$GET$.<clinit>(Method.scala:55)
	... 2 more

--------------------------------------------------------------------------------
timestamp=2022-11-16T18:37:56.009881Z level=ERROR thread=#zio-fiber-0 message="" cause="Exception in thread "zio-fiber-4" java.lang.ExceptionInInitializerError: null
	at izumi.reflect.macrortti.LightTypeTag$$anon$8.unpickle(LightTypeTag.scala:490)
	at izumi.reflect.macrortti.LightTypeTag$$anon$8.unpickle(LightTypeTag.scala:487)
	at izumi.reflect.thirdparty.internal.boopickle.CompositePickler.unpickle(CompositePicklers.scala:56)
	at izumi.reflect.macrortti.LightTypeTag$.deserializeRefString(LightTypeTag.scala:395)
	at izumi.reflect.macrortti.LightTypeTag$ParsedLightTypeTag210.ref(LightTypeTag.scala:377)
	at izumi.reflect.macrortti.LightTypeTag.decompose(LightTypeTag.scala:69)
	at zio.VersionSpecific.taggedGetServices(VersionSpecific.scala:75)
	at zio.VersionSpecific.taggedGetServices$(VersionSpecific.scala:26)
	at zio.package$.taggedGetServices(package.scala:23)
	at zio.ZEnvironment.prune(ZEnvironment.scala:65)
	at zio.ZEnvironment.union(ZEnvironment.scala:95)
	at zio.Scope$ExtendPartiallyApplied$.apply$extension$$anonfun$1(Scope.scala:172)
	at zio.ZIO.provideSomeEnvironment$$anonfun$1$$anonfun$1(ZIO.scala:1240)
	at zio.ZIO.provideEnvironment(ZIO.scala:1220)
	at zio.ZIO.provideEnvironment$(ZIO.scala:52)
	at zio.ZIO$OnSuccess.provideEnvironment(ZIO.scala:5533)
	at zio.ZIO.provideSomeEnvironment$$anonfun$1(ZIO.scala:1240)
	Suppressed: java.lang.NoSuchFieldException: 0bitmap$5
		at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
		at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
		at izumi.reflect.macrortti.LightTypeTagRef$NameReference.<clinit>(LightTypeTagRef.scala:343)
		at izumi.reflect.macrortti.LightTypeTag$$anon$8.unpickle(LightTypeTag.scala:490)
		at izumi.reflect.macrortti.LightTypeTag$$anon$8.unpickle(LightTypeTag.scala:487)
		at izumi.reflect.thirdparty.internal.boopickle.CompositePickler.unpickle(CompositePicklers.scala:56)
		at izumi.reflect.macrortti.LightTypeTag$.deserializeRefString(LightTypeTag.scala:395)
		at izumi.reflect.macrortti.LightTypeTag$ParsedLightTypeTag210.ref(LightTypeTag.scala:377)
		at izumi.reflect.macrortti.LightTypeTag.decompose(LightTypeTag.scala:69)
		at zio.VersionSpecific.taggedGetServices(VersionSpecific.scala:75)
		at zio.VersionSpecific.taggedGetServices$(VersionSpecific.scala:26)
		at zio.package$.taggedGetServices(package.scala:23)
		at zio.ZEnvironment.prune(ZEnvironment.scala:65)
		at zio.ZEnvironment.union(ZEnvironment.scala:95)
		at zio.Scope$ExtendPartiallyApplied$.apply$extension$$anonfun$1(Scope.scala:172)
		at zio.ZIO.provideSomeEnvironment$$anonfun$1$$anonfun$1(ZIO.scala:1240)
		at zio.ZIO.provideEnvironment(ZIO.scala:1220)
		at zio.ZIO.provideEnvironment$(ZIO.scala:52)
		at zio.ZIO$OnSuccess.provideEnvironment(ZIO.scala:5533)
		at zio.ZIO.provideSomeEnvironment$$anonfun$1(ZIO.scala:1240)"

--------------------------------------------------------------------------------
timestamp=2022-11-16T18:43:25.741961Z level=ERROR thread=#zio-fiber-0 message="" cause="Exception in thread "zio-fiber-4" java.lang.ExceptionInInitializerError: null
	at izumi.reflect.internal.fundamentals.platform.console.TrivialLogger$Config$.console(TrivialLogger.scala:71)
	at izumi.reflect.macrortti.LightTypeTagInheritance.isChild(LightTypeTagInheritance.scala:61)
	at izumi.reflect.macrortti.LightTypeTag.$less$colon$less(LightTypeTag.scala:61)
	at zio.VersionSpecific.taggedIsSubtype$$anonfun$1(VersionSpecific.scala:63)
	at java.base@17.0.5/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at zio.VersionSpecific.taggedIsSubtype(VersionSpecific.scala:63)
	at zio.VersionSpecific.taggedIsSubtype$(VersionSpecific.scala:26)
	at zio.package$.taggedIsSubtype(package.scala:23)
	at zio.ZEnvironment.$anonfun$1$$anonfun$1(ZEnvironment.scala:68)
	at scala.collection.IterableOnceOps.exists(IterableOnce.scala:603)
	at scala.collection.IterableOnceOps.exists$(IterableOnce.scala:600)
	at scala.collection.AbstractIterable.exists(Iterable.scala:933)
	at zio.ZEnvironment.$anonfun$1(ZEnvironment.scala:68)
	at scala.collection.immutable.Set$Set1.filterImpl(Set.scala:172)
	at scala.collection.immutable.Set$Set1.filterImpl(Set.scala:156)
	at scala.collection.StrictOptimizedIterableOps.filterNot(StrictOptimizedIterableOps.scala:220)
	at scala.collection.StrictOptimizedIterableOps.filterNot$(StrictOptimizedIterableOps.scala:220)
	at scala.collection.immutable.Set$Set1.filterNot(Set.scala:156)
	at zio.ZEnvironment.prune(ZEnvironment.scala:68)
	at zio.ZEnvironment.union(ZEnvironment.scala:95)
	at zio.Scope$ExtendPartiallyApplied$.apply$extension$$anonfun$1(Scope.scala:172)
	at zio.ZIO.provideSomeEnvironment$$anonfun$1$$anonfun$1(ZIO.scala:1240)
	at zio.ZIO.provideEnvironment(ZIO.scala:1220)
	at zio.ZIO.provideEnvironment$(ZIO.scala:52)
	at zio.ZIO$OnSuccess.provideEnvironment(ZIO.scala:5533)
	at zio.ZIO.provideSomeEnvironment$$anonfun$1(ZIO.scala:1240)
	Suppressed: java.lang.NoSuchFieldException: 0bitmap$1
		at java.base@17.0.5/java.lang.Class.getDeclaredField(DynamicHub.java:961)
		at scala.runtime.LazyVals$.getOffset(LazyVals.scala:138)
		at izumi.reflect.internal.fundamentals.platform.console.TrivialLogger$Config.<clinit>(TrivialLogger.scala:71)
		at izumi.reflect.internal.fundamentals.platform.console.TrivialLogger$Config$.console(TrivialLogger.scala:71)
		at izumi.reflect.macrortti.LightTypeTagInheritance.isChild(LightTypeTagInheritance.scala:61)
		at izumi.reflect.macrortti.LightTypeTag.$less$colon$less(LightTypeTag.scala:61)
		at zio.VersionSpecific.taggedIsSubtype$$anonfun$1(VersionSpecific.scala:63)
		at java.base@17.0.5/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
		at zio.VersionSpecific.taggedIsSubtype(VersionSpecific.scala:63)
		at zio.VersionSpecific.taggedIsSubtype$(VersionSpecific.scala:26)
		at zio.package$.taggedIsSubtype(package.scala:23)
		at zio.ZEnvironment.$anonfun$1$$anonfun$1(ZEnvironment.scala:68)
		at scala.collection.IterableOnceOps.exists(IterableOnce.scala:603)
		at scala.collection.IterableOnceOps.exists$(IterableOnce.scala:600)
		at scala.collection.AbstractIterable.exists(Iterable.scala:933)
		at zio.ZEnvironment.$anonfun$1(ZEnvironment.scala:68)
		at scala.collection.immutable.Set$Set1.filterImpl(Set.scala:172)
		at scala.collection.immutable.Set$Set1.filterImpl(Set.scala:156)
		at scala.collection.StrictOptimizedIterableOps.filterNot(StrictOptimizedIterableOps.scala:220)
		at scala.collection.StrictOptimizedIterableOps.filterNot$(StrictOptimizedIterableOps.scala:220)
		at scala.collection.immutable.Set$Set1.filterNot(Set.scala:156)
		at zio.ZEnvironment.prune(ZEnvironment.scala:68)
		at zio.ZEnvironment.union(ZEnvironment.scala:95)
		at zio.Scope$ExtendPartiallyApplied$.apply$extension$$anonfun$1(Scope.scala:172)
		at zio.ZIO.provideSomeEnvironment$$anonfun$1$$anonfun$1(ZIO.scala:1240)
		at zio.ZIO.provideEnvironment(ZIO.scala:1220)
		at zio.ZIO.provideEnvironment$(ZIO.scala:52)
		at zio.ZIO$OnSuccess.provideEnvironment(ZIO.scala:5533)
		at zio.ZIO.provideSomeEnvironment$$anonfun$1(ZIO.scala:1240)"
```

The project I've been testing on is at https://github.com/carlosedp/zio-scalajs-stack and it's configured to build with GraalVM and Scala 2.13 here https://github.com/carlosedp/zio-scalajs-stack/blob/main/build.sc#L57. I build with `./mill show backend.nativeImage`. 

Let me know if I can help by providing any additional info.
Hope it helps.

--- a/library/src/scala/runtime/LazyVals.scala
+++ b/library/src/scala/runtime/LazyVals.scala
@@ -10,13 +10,19 @@ import scala.annotation.*
 object LazyVals {
   @nowarn
   private[this] val unsafe: sun.misc.Unsafe = {
-    val unsafeField = classOf[sun.misc.Unsafe].getDeclaredField("theUnsafe").nn
-    if unsafeField.getType == classOf[sun.misc.Unsafe] then
-      unsafeField.setAccessible(true)
-    else throw new ExceptionInInitializerError {
-      new IllegalStateException("Can't find instance of sun.misc.Unsafe")
-    }
-    unsafeField.get(null).asInstanceOf[sun.misc.Unsafe]
+    def throwInitializationException() = 
+      throw new ExceptionInInitializerError(
+        new IllegalStateException("Can't find instance of sun.misc.Unsafe")
+      )
+    try
+      val unsafeField = classOf[sun.misc.Unsafe].getDeclaredField("theUnsafe").nn
+      if unsafeField.getType == classOf[sun.misc.Unsafe] then
+        unsafeField.setAccessible(true)
+        unsafeField.get(null).asInstanceOf[sun.misc.Unsafe]
+      else
+        throwInitializationException()
+    catch case _: NoSuchFieldException =>
+      throwInitializationException()
   }
 
   private[this] val base: Int = {

--- a/library/src/scala/runtime/LazyVals.scala
+++ b/library/src/scala/runtime/LazyVals.scala
@@ -9,19 +9,15 @@ import scala.annotation.*
  */
 object LazyVals {
   @nowarn
-  private[this] val unsafe: sun.misc.Unsafe =
-      classOf[sun.misc.Unsafe].getDeclaredFields.nn.find { field =>
-        field.nn.getType == classOf[sun.misc.Unsafe] && {
-          field.nn.setAccessible(true)
-          true
-        }
-      }
-      .map(_.nn.get(null).asInstanceOf[sun.misc.Unsafe])
-      .getOrElse {
-        throw new ExceptionInInitializerError {
-          new IllegalStateException("Can't find instance of sun.misc.Unsafe")
-        }
-      }
+  private[this] val unsafe: sun.misc.Unsafe = {
+    val unsafeField = classOf[sun.misc.Unsafe].getDeclaredField("theUnsafe").nn
+    if unsafeField.getType == classOf[sun.misc.Unsafe] then
+      unsafeField.setAccessible(true)
+    else throw new ExceptionInInitializerError {
+      new IllegalStateException("Can't find instance of sun.misc.Unsafe")
+    }
+    unsafeField.get(null).asInstanceOf[sun.misc.Unsafe]
+  }
 
   private[this] val base: Int = {
     val processors = java.lang.Runtime.getRuntime.nn.availableProcessors()

